xds/internal/xdsclient/transport/transport.go

```diff
@@ -0,0 +1,564 @@
+// user, while the version and nonce are maintained internally.
+//
+// The resources map, which keeps track of the resources being requested, is
+// updated here. Any subsequent stream failure will re-request resources stored
+// in this map.
+//
+// Returns the list of resources, resource type url, version and nonce.
+func (t *Transport) processResourceRequest(req *resourceRequest) ([]string, string, string, string) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	resources := sliceToMap(req.resources)
+	t.resources[req.url] = resources
+	return req.resources, req.url, t.versions[req.url], t.nonces[req.url]
+}
+
+type ackRequest struct {
+	url     string // Resource type URL.
+	version string // NACK if version is an empty string.
+	nonce   string
+	errMsg  string // Empty unless it's a NACK.
+	// ACK/NACK are tagged with the stream it's for. When the stream is down,
+	// all the ACK/NACK for this stream will be dropped, and the version/nonce
+	// won't be updated.
+	stream grpc.ClientStream
+}
+
+// processAckRequest pulls the fields needed to send out an ADS ACK. The nonces
+// and versions map is updated.
+//
+// Returns the list of resources, resource type url, version, nonce, and an
+// indication of whether an ACK should be sent on the wire or not.
+func (t *Transport) processAckRequest(ack *ackRequest, stream grpc.ClientStream) ([]string, string, string, string, bool) {
+	if ack.stream != stream {
+		// If ACK's stream isn't the current sending stream, this means the ACK
+		// was pushed to queue before the old stream broke, and a new stream has
+		// been started since. Return immediately here so we don't update the
+		// nonce for the new stream.
+		return nil, "", "", "", false
+	}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// Update the nonce irrespective of whether we send the ACK request on wire.
+	// An up-to-date nonce is required for the next request.
+	nonce := ack.nonce
+	t.nonces[ack.url] = nonce
+
+	s, ok := t.resources[ack.url]
+	if !ok || len(s) == 0 {
+		// We don't send the ACK request if there are no resources of this type
+		// in our resources map. This can be either when the server sends
+		// responses before any request, or the resources are removed while the
+		// ackRequest was in queue). If we send a request with an empty
+		// resource name list, the server may treat it as a wild card and send
+		// us everything.
+		return nil, "", "", "", false
+	}
+	resources := mapToSlice(s)
+
+	version := ack.version
+	if version == "" {
+		// This is a NACK. Get the previously ACKed version, which could also be
+		// an empty string.  This can happen if there wasn't any ACK before.
+		version = t.versions[ack.url]
+	} else {
+		// This is an ACK. Update the versions map.
+		t.versions[ack.url] = version
+	}
+	return resources, ack.url, version, nonce, true
+}
+
+// Close closes the Transport and frees any associated resources.
+func (t *Transport) Close() {
+	t.stopRunGoroutine()
```

Comment:
I think in principle this should block until the stream has exited. This may be worth doing now.

Reply:
Done.